aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,375 @@
+  int dimw = 2;
+  int numBatch = 1;
+
+  int ndims = input.ndimension();
+  for (int i = 0; i < ndims; i++) {
+     AT_CHECK(input.size(i) > 0,
+       "fractional_max_pool2d(): expected input to have non-empty spatial dimensions, "
+       "but input has sizes ", input.sizes(), " with dimension ", i, " being "
+       "empty");
+   }
+
+   AT_CHECK((ndims == 3 || ndims == 4),
+     "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  if (ndims == 4) {
+    numBatch = input.size(0);
+    planeDim++;
+    dimh++;
+    dimw++;
+  }
+
+  /* sizes */
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(dimh);
+  int inputW = input.size(dimw);
+
+  int outputH = output_size[0];
+  int outputW = output_size[1];
+  int poolSizeH = pool_size[0];
+  int poolSizeW = pool_size[1];
+
+  AT_CHECK(outputH + poolSizeH - 1 <= inputH,
+             "fractional_max_pool2d(): pool_size height ", poolSizeH,
+             " too large relative to input height ", inputH);
+  AT_CHECK(outputW + poolSizeW - 1 <= inputW,
+           "pool_size width ", poolSizeW,
+           " too large relative to input width ", inputW);
+
+  if (ndims == 3) {
+    /* resize output */
+    output.resize_({numPlanes, outputH, outputW});
+    /* indices will contain the locations for each output point */
+    indices.resize_({numPlanes, outputH, outputW});
+  } else {
+    output.resize_({numBatch, numPlanes, outputH, outputW});
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+  }
+
+  auto output_ = output;
+  auto input_ = input;
+  auto indices_ = indices;
+
+  if(ndims == 3) {
+    output_.reshape({1, numPlanes, outputH, outputW});
+    indices_.reshape({1, numPlanes, outputH, outputW});
+    input_.reshape({1, input.size(0), input.size(1), input.size(2)});
+  }
+
+  // block is limited to 4 warps
+  // grid handles overflow per each plane
+  int outputPlaneSize = output_.size(2) *
+    output_.size(3);
+  dim3 grid((outputPlaneSize + 127) / 128,
+            input_.size(1),
+            input_.size(0));
+  dim3 block(outputPlaneSize > 128 ? 128 : outputPlaneSize);
+
+#define FMP(POOL_W) \
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.type(), \
+    "fractional_max_pool2d_out_cuda_frame", \
+    [&] { \
+      auto devInput = input_.packed_accessor<scalar_t, 4>(); \
+      auto devOutput = output_.packed_accessor<scalar_t, 4>();\
+      auto devIndices = indices_.packed_accessor<int64_t, 4>();\
+      auto devSamples = randomSamples.packed_accessor<scalar_t, 3>();\
+      fractional_max_pool2d_out_cuda_frame<scalar_t, POOL_W>\
+        <<<grid, block, 0, at::cuda::getCurrentCUDAStream()>>>(\
+          devOutput, devIndices, devInput, devSamples,\
+          poolSizeH, poolSizeW);\
+       }\
+     )
+
+#define FMP_CASE(POOL_W) case POOL_W: FMP(POOL_W); break
+
+    switch (poolSizeW) {
+      FMP_CASE(2);
+      FMP_CASE(3);
+      FMP_CASE(4);
+      FMP_CASE(5);
+      FMP_CASE(6);
+      FMP_CASE(7);
+      default: FMP(-1);
+    }
```

Comment:
you lost the check for `cudaGetLastError` here

Reply:
done